### DISCUSSION TITLE
Boundary condition:NeumannBC did not working

### DISCUSSION THREAD

**Discussion by emmelines** [idaholab/moose#18608] [Q&A Modules: Phase field] [answered]
Hi 

I am using the phase-field model to simulate two phase-separated microstructure evolution. In my model, I set NeumannBC as the boundary condition. However, when I check the boundary flux using SideAverageValue, the result showed the value of flux is not zero. Is there any reason for the boundary condition cannot work properly? 

Thank you.

**Comment by GiudGiud** [comment]
Hello

SideAverageValue computes the average of a variable not really of a flux. Are you using the gradient as the variable?
If not you may be wanting to use SideDiffusiveFluxIntegral

Guillaume

**Reply by emmelines** [reply]
Hi

I did use gradient variable, so the SideAverageValue should fit in my case. Is it?
Thanks.

Emmeline

**Reply by GiudGiud** [reply]
But a SideAverageValue does not take gradient variables? Did you use the gradient component as the variable?
I think pasting your variable block and the postprocessor block would help here

**Reply by emmelines** [reply]
Hi 

The following are my variable and postprocessor. Thank you for staying on this problem with me. I appreciate. 

```
[Variables]
  [./c]
  [../]
  [./w]
  [../]
  [./eta_a1]
  [../]
  [./eta_b1]
  [../]
  [./eta_b2]
  [../]
  [./eta_b3]
  [../]
  [./eta_b4]
  [../]
[]

[Postprocessors]
  [./ElementInt_c]
    type = ElementIntegralVariablePostprocessor
    variable = c
  [../]
  [./precip_area] # since eta=1.0 inside the precipitate
    type = ElementIntegralVariablePostprocessor
    variable = 'eta_a1 eta_b1 eta_b2 eta_b3 eta_b4'
  [../]
  [./total_F]
    type = ElementIntegralVariablePostprocessor
    variable = total_F
  [../]
  [./F]
    type = ElementIntegralVariablePostprocessor
    variable = F
  [../]
  [./dwdx]
    type = ElementIntegralVariablePostprocessor
    variable = dwdx
    execute_on = 'initial timestep_end'
  [../]
  [./dwdy]
    type = ElementIntegralVariablePostprocessor
    variable = dwdy
    execute_on = 'initial timestep_end'
  [../]
#######
  [./total_Fc]
  type = ElementIntegralVariablePostprocessor
  variable = total_Fc
  [../]
  [./total_Feta]
  type = ElementIntegralVariablePostprocessor
  variable = total_Feta
  [../]
  [./F_c]
    type = ElementIntegralVariablePostprocessor
    variable = F_c
    execute_on = 'initial timestep_end'
  [../]
  [./F_eta]
    type = ElementIntegralVariablePostprocessor
    variable = F_eta
    execute_on = 'initial timestep_end'
  [../]
  [./F_etaandc]
    type = ElementIntegralVariablePostprocessor
    variable = F_etaandc
    execute_on = 'initial timestep_end'
  [../]
  [./right_jx]
    type = SideAverageValue
    variable = jx
    boundary = right
  [../]
  [./right_jy]
    type = SideAverageValue
    variable = jy
    boundary = right
  [../]
  [./left_jx]
    type = SideAverageValue
    variable = jx
    boundary = left
  [../]
  [./left_jy]
    type = SideAverageValue
    variable = jy
    boundary = left
  [../]
  [./right_j_tot]
    type = SideAverageValue
    variable = j_tot
    boundary = right
  [../]
  [./left_j_tot]
    type = SideAverageValue
    variable = j_tot
    boundary = left
  [../]
  [./top_jx]
    type = SideAverageValue
    variable = jx
    boundary = top
  [../]
  [./top_jy]
    type = SideAverageValue
    variable = jy
    boundary = top
  [../]
  [./bottom_jx]
    type = SideAverageValue
    variable = jx
    boundary = bottom
  [../]
  [./bottom_jy]
    type = SideAverageValue
    variable = jy
    boundary = bottom
  [../]
  [./top_j_tot]
    type = SideAverageValue
    variable = j_tot
    boundary = top
  [../]
  [./bottom_j_tot]
    type = SideAverageValue
    variable = j_tot
    boundary = bottom
  [../]
```

**Reply by GiudGiud** [reply]
So jx and jy are AuxVariables? Did you define them in the input file ?
Tagging @laagesen in case there s something special about phase field here

**Reply by emmelines** [reply]
I defined jx and jy in AuxKernels. 
```
[AuxKernels]
[./dwdx]
    type = VariableGradientComponent
    variable = dwdx
    gradient_variable = w
    component = x
  [../]
  [./dwdy]
    type = VariableGradientComponent
    variable = dwdy
    gradient_variable = w
    component = y
  [../]
  [./jx]
    type = ParsedAux
    variable = jx
    args = 'dwdx'
    function = '-1.0*dwdx'
  [../]
  [./jy]
    type = ParsedAux
    variable = jy
    args = 'dwdy'
    function = '-1.0*dwdy'
  [../]

**Reply by GiudGiud** [reply]
Not sure. While we wait for someone else to pitch in, you are dealing with a diffusive flux right?
If so, SideDiffusiveFluxIntegral will compute it from the variable definition. Does that work for you?

**Reply by emmelines** [reply]
Hi 

I am trying the SideDiffusiveFluxIntegral, and it seems like my library has to update first. I will get back to you once I complete the computation. Thanks a lot!

**Reply by dschwen** [reply]
```
  [./dwdx]
    type = ElementIntegralVariablePostprocessor
    variable = dwdx
    execute_on = 'initial timestep_end'
  [../]
  [./dwdy]
    type = ElementIntegralVariablePostprocessor
    variable = dwdy
    execute_on = 'initial timestep_end'
  [../]
```

You're integrating the gradient components over a volume rather than evaluating it on the side where you apply the BC.

**Reply by GiudGiud** [reply]
they are talking about these postprocessors I think
```  [./top_jx]
    type = SideAverageValue
    variable = jx
    boundary = top
  [../]
  [./top_jy]
    type = SideAverageValue
    variable = jy
    boundary = top
  [../]
  [./bottom_jx]
    type = SideAverageValue
    variable = jx
    boundary = bottom
  [../]
  [./bottom_jy]
    type = SideAverageValue
    variable = jy
    boundary = bottom
  [../]
```

**Comment by jiangwen84** [comment, marked as answer]
@emmelines In FEM, NeumannBC is weakly enforced. The flux won't be exactly zero. However, the flux will become more accurate as mesh is refined. 

For your problem, you can take a look at the global concentration. Zero flux will result in conserved concentration.

**Reply by emmelines** [reply]
@jiangwen84 
Could I ask what is the reason for NeumannBC has weak enforcement in FEM?
Thank you.

**Reply by jiangwen84** [reply]
You can go to https://mooseframework.inl.gov/workshop/workshop/index.html#/2/7 

The third term of the weak form is the Neumann BC. As you can see, that term is weighted by the test function \phi_i. That means this term is weakly enforced. As you increase the test function space (use higher order and/or small mesh size), the flux k\grad T n will become more accurate.
